third_party/amd/lib/TritonAMDGPUToLLVM/ReduceOpToLLVM.cpp

```diff
@@ -17,6 +17,129 @@ using ::mlir::triton::gpu::getTotalElemsPerThread;
 
 namespace AMD {
 namespace {
+struct ReduceOpPromotionConversion
+    : public ConvertTritonGPUReduceScanToLLVMPattern<triton::ReduceOp> {
+public:
+  ReduceOpPromotionConversion(
+      TritonGPUToLLVMTypeConverter &typeConverter, ModuleAllocation &allocation,
+      ConvertTritonGPUOpToLLVMPatternBase::IndexCacheInfo &indexCacheInfo,
+      int computeCapability, PatternBenefit benefit)
+      : ConvertTritonGPUReduceScanToLLVMPattern<triton::ReduceOp>(
+            typeConverter, allocation, indexCacheInfo, benefit),
+        computeCapability(computeCapability) {}
+
+  LogicalResult
+  matchAndRewrite(triton::ReduceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    
+    // promote op
+    rewriter.modifyOpInPlace(op, [&]() {
+      // promote operands
+      SmallVector<Value> newOperands;
+      for (OpOperand &operand : op->getOpOperands()) {
+        auto val = operand.get();
+        auto oldType = val.getType().cast<RankedTensorType>();
+        auto elemType = oldType.getElementType();
+        if (elemType.isInteger(16) || elemType.isInteger(8)) {
+          auto newType = oldType.cloneWith(std::nullopt, i32_ty);
+          auto promotedVal =
+              rewriter.create<mlir::arith::ExtSIOp>(op->getLoc(), newType, val);
+          newOperands.push_back(promotedVal);
+        } else if (elemType.isF16()) {
+          auto newType = oldType.cloneWith(std::nullopt, f32_ty);
+          auto promotedVal =
+              rewriter.create<mlir::arith::ExtFOp>(op->getLoc(), newType, val);
+          newOperands.push_back(promotedVal);
+        } else {
+          newOperands.push_back(val);
+        }
+      }
+      op->setOperands(newOperands);
+
+      // promote results
+      for (Value result : op.getResults()) {
+        auto type = result.getType();
+        if (type.isInteger(16) || type.isInteger(8)) {
+          result.setType(i32_ty);
+        } else if (type.isF16()) {
+          result.setType(f32_ty);
+
+          // add trunc if float result type was changed
```

Comment:
Why do we need to truncate the result back for float but not for integers?

Reply:
It was not needed for the ints to pass the tests but I can explicitly add one.

Reply:
added ones